core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -0,0 +1,390 @@
+    struct.set(VersionField, PidSnapshotVersion)
+    struct.set(CrcField, 0L) // we'll fill this after writing the entries
+    val entriesArray = entries.map {
+      case (pid, entry) =>
+        val pidEntryStruct = struct.instance(PidEntriesField)
+        pidEntryStruct.set(PidField, pid)
+          .set(EpochField, entry.epoch)
+          .set(LastSequenceField, entry.lastSeq)
+          .set(LastOffsetField, entry.lastOffset)
+          .set(NumRecordsField, entry.numRecords)
+          .set(TimestampField, entry.timestamp)
+        pidEntryStruct
+    }.toArray
+    struct.set(PidEntriesField, entriesArray)
+
+    val buffer = ByteBuffer.allocate(struct.sizeOf)
+    struct.writeTo(buffer)
+    buffer.flip()
+
+    // now fill in the CRC
+    val crc = Crc32C.compute(buffer, PidEntriesOffset, buffer.limit - PidEntriesOffset)
+    ByteUtils.writeUnsignedInt(buffer, CrcOffset, crc)
+
+    val fos = new FileOutputStream(file)
+    try {
+      fos.write(buffer.array, buffer.arrayOffset, buffer.limit)
+    } finally {
+      fos.close()
+    }
+  }
+
+  private def verifyFileName(name: String): Boolean = FilenamePattern.findFirstIn(name).isDefined
+
+  private def offsetFromFile(file: File): Long = {
+    s"${file.getName.replace(s".$FilenameSuffix", "")}".toLong
+  }
+
+  private def formatFileName(lastOffset: Long): String = {
+    // The files will be named '$lastOffset.snapshot' and located in 'logDir/pid-mapping'
+    s"$lastOffset.$FilenameSuffix"
+  }
+
+}
+
+/**
+ * Maintains a mapping from ProducerIds (PIDs) to metadata about the last appended entries (e.g.
+ * epoch, sequence number, last offset, etc.)
+ *
+ * The sequence number is the last number successfully appended to the partition for the given identifier.
+ * The epoch is used for fencing against zombie writers. The offset is the one of the last successful message
+ * appended to the partition.
+ */
+@nonthreadsafe
+class ProducerIdMapping(val config: LogConfig,
+                        val topicPartition: TopicPartition,
+                        val snapParentDir: File,
+                        val maxPidExpirationMs: Int) extends Logging {
+  import ProducerIdMapping._
+
+  val snapDir: File = new File(snapParentDir, DirnamePrefix)
+  snapDir.mkdir()
+
+  private val pidMap = mutable.Map[Long, ProducerIdEntry]()
+  private var lastMapOffset = 0L
+  private var lastSnapOffset = 0L
+
+  /**
+   * Returns the last offset of this map
+   */
+  def mapEndOffset = lastMapOffset
+
+  /**
+   * Get a copy of the active producers
+   */
+  def activePids: immutable.Map[Long, ProducerIdEntry] = pidMap.toMap
+
+  /**
+   * Load a snapshot of the id mapping or return empty maps
+   * in the case the snapshot doesn't exist (first time).
+   */
+  private def loadFromSnapshot(logEndOffset: Long, checkNotExpired:(ProducerIdEntry) => Boolean) {
+    pidMap.clear()
+
+    var loaded = false
+    while (!loaded) {
+      lastSnapshotFile(logEndOffset) match {
```

Comment:
Not sure if it would be more clear, but probably more efficient. I made the change this way so that we retain the existing logic for picking the latest snapshot less than the given offset.